shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -993,37 +995,158 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, var ne
-                // assert(!mut)
-                val fprov = tp(a.toLoc, "argument")
-                val tym = typeArg(a)
-                (n, tym.toUpper(fprov))
-              })(as match { // TODO dedup w/ general Tup case
-                case Nil | ((N, _) :: Nil) => noProv
-                case _ => tp(tup.toLoc, "argument list")
-              })
-            case _ => // can happen in the old parser
-              typeArg(a)
+        println("applying-to-function => " + f + "<" + f.getClass() + ">" + a + "<" + a.getClass() + ">")
+        println("now we should desugar in case of named args!")
+        /////////////////////////////
+        def typeUnnamedApp(f: Term, a: Term) = {
+          val f_ty = typeMonomorphicTerm(f)
+          val a_ty = {
+            def typeArg(a: Term): ST =
+              if (!generalizeArguments) typePolymorphicTerm(a)
+              else ctx.poly { implicit ctx => typePolymorphicTerm(a) }
+            a match {
+              case tup @ Tup(as) =>
+                TupleType(as.map { case (n, Fld(mut, spec, a)) => // TODO handle mut?
+                  // assert(!mut)
+                  val fprov = tp(a.toLoc, "argument")
+                  val tym = typeArg(a)
+                  (n, tym.toUpper(fprov))
+                })(as match { // TODO dedup w/ general Tup case
+                  case Nil | ((N, _) :: Nil) => noProv
+                  case _ => tp(tup.toLoc, "argument list")
+                })
+              case _ => // can happen in the old parser
+                typeArg(a)
+            }
+          }
+          println("f and f_ty => " + Helpers.inspect(f) + " " + f_ty)
+          println("a and a_ty => " + Helpers.inspect(a) + " " + a_ty)
+          val res = freshVar(prov, N)
+          val arg_ty = mkProxy(a_ty, tp(a.toCoveringLoc, "argument"))
+            // ^ Note: this no longer really makes a difference, due to tupled arguments by default
+          val funProv = tp(f.toCoveringLoc, "applied expression")
+          val fun_ty = mkProxy(f_ty, funProv)
+            // ^ This is mostly not useful, except in test Tuples.fun with `(1, true, "hey").2`
+          val resTy = con(fun_ty, FunctionType(arg_ty, res)(
+            prov
+            // funProv // TODO: better?
+            ), res)
+          resTy
+        }
+        /////////////////////////////
+        def desugarNamedArgs(term: Term, f: Var, a: Tup): SimpleType = {
+          val Var(name) = f;
+          ctx.get(name) match {
+            case Some(value) => 
+              println("value => " + value  + " " + value.getClass())
+              value match {
+                case VarSymbol(ty, definingVar) => 
+                  println("definingVar => " + definingVar)
+                  println("ty => " + ty + " " + ty.getClass())
+                  ty match {
+                    case FunctionType(lhs, rhs) => 
+                      println("lhs, rhs => " + lhs + " " + rhs + " " + lhs.getClass() + " " + rhs.getClass())
```

Comment:
```suggestion
                      println(s"lhs, rhs => $lhs $rhs ${lhs.getClass()} ${rhs.getClass()}")
```
You can use string templates to avoid unnecessary string concat.